### ISSUE TITLE
Connection pattern: fully connected and adressable banks

### ISSUE THREAD

**Issue by cmnrd** [icyphy/lingua-franca#415]
I noticed a pattern that is quite common in the Savina benchmarks suite, but that is not supported by our current connection syntax.  I am not sure if the name in the title captures well what this is about, so let me explain.

Building a fully connected network of reactors in a bank where each instance broadcasts its messages to all other instances in the bank is quite simple. For instance, the following program
```
target Cpp {
    threads: 1
}

reactor Node(bank_index: size_t(0), num_nodes: size_t(4)) {
    input[num_nodes] in: size_t
    output out: size_t
    
    reaction (startup) -> out{=
     	std::cout << "Hello from node " << bank_index << "!\n";
     	// broadcast my ID to everyone
     	out.set(bank_index);    
    =}
    
    reaction (in) {=
        std::cout << "Node " << bank_index << " received messages from ";
     	for (auto& port : in) {
            if (port.is_present()) {
                std::cout << *port.get() << ", ";
            }
        }   
        std::cout << '\n';
    =}
}

main reactor(num_nodes: size_t(4)) {
    nodes = new[num_nodes] Node(num_nodes=num_nodes);
    (nodes.out)+ -> nodes.in;
}
```
will print
``` 
Hello from node 3!
Hello from node 2!
Hello from node 1!
Hello from node 0!
Node 3 received messages from 0, 1, 2, 3, 
Node 2 received messages from 0, 1, 2, 3, 
Node 1 received messages from 0, 1, 2, 3, 
Node 0 received messages from 0, 1, 2, 3, 
```
All the magic lies in the line `(nodes.out)+ -> nodes.in;` which broadcasts the output of each instance, to the inputs of all instances.

So far, so good. But what if we want to make this a little more sophisticated and make the instances addressable, so that the sender does not broadcast to all receivers but can send messages directly to a specific receiver? Naively, we could try the following:
```
target Cpp {
    threads: 1
}

reactor Node(bank_index: size_t(0), num_nodes: size_t(4)) {
    input[num_nodes] in: size_t
    output[num_nodes] out: size_t
    
    reaction (startup) -> out{=
     	std::cout << "Hello from node " << bank_index << "!\n";
     	// send my ID only to my right neighbour
     	out[(bank_index + 1) % num_nodes].set(bank_index);    
    =}
    
    reaction (in) {=
        std::cout << "Node " << bank_index << " received messages from ";
     	for (auto& port : in) {
            if (port.is_present()) {
                std::cout << *port.get() << ", ";
            }
        }   
        std::cout << '\n';
    =}
}

main reactor(num_nodes: size_t(4)) {
    nodes = new[num_nodes] Node(num_nodes=num_nodes);
    nodes.out -> nodes.in;
}
```
In this program, each node is supposed to send a message to its right neighbor (higher `bank_index`). However, the output produced by the program is
```
Hello from node 3!
Hello from node 2!
Hello from node 1!
Hello from node 0!
Node 3 received messages from 3, 
Node 2 received messages from 2, 
Node 1 received messages from 1, 
Node 0 received messages from 0, 
```
This indicates that the nodes actually only send messages to themselves. Considering what `nodes.out -> nodes.in;` does, this is not all too surprising. Expressed in C++ pseudocode, the connection will be created using nested for loops similar to this:
```
for (size_t i = 0; i < num_nodes; i++) {
    for (size_t j = 0; j < num_nodes; j++) {
        connect(nodes[i].out[j], nodes[i].in[j]);
    }
}
```
However, what we actually want is this (notice the swapped indices):
```
for (size_t i = 0; i < num_nodes; i++) {
    for (size_t j = 0; j < num_nodes; j++) {
        connect(nodes[i].out[j], nodes[j].in[i]);
    }
}
```
Now my question: How can we distinguish the two different connection patterns in our syntax? I believe that we want to support both patterns, which means that they should be distinguishable somehow. Maybe we could use `-x->` for "cross connection"?

**Comment by edwardalee** [comment]
Interesting observation. One way to think of the connection in the second example:
```
    nodes.out -> nodes.in;
```
is this: There are 16 channels on the left and 16 on the right (bank of 4, multiport of width 4). The connection logic either iterates over banks first, followed by multiports, or the other way around. I don't recall which it is, but it doesn't matter because both sides iterate in the same order. The [language specification](https://github.com/icyphy/lingua-franca/wiki/Multiports-and-Banks-of-Reactors) does not say what order is used, but it probably should.  Then your proposed syntax `-x->` would reverse the order, but again we would need to specify which side iterates which way.  An alternative syntax might be to actually specify the iteration order on both sides.  E.g.,
```
    (nodes.out)b -> (nodes.in)m;
```
means to iterate over banks first on the left and multiports first on the right. Do we need this additional generality?

Also, would it make sense to be able to combine this syntax with `(...)+`? In your first example, I think you would get the same connection pattern if you say
```
    (nodes.out)+ -> (nodes.in)b;
```
or
```
    (nodes.out)+ -> (nodes.in)m;
```
However, if the output is also a multiport, say of width 2, then I think these will yield different connections (?).

**Comment by cmnrd** [comment]
Yes, I think an order modifier like you propose would also work. But, I don't know if we need this generality, and I am a bit afraid that this would add a whole new layer of complexity. What would `((a.out)m, (b.out)b)+ -> (c.in)b, (d.in)m, (e.in)b` do? I am having a hard time wrapping my head around more complex connection pattern already, this would just add another dimension to the problem.

I guess we can try to keep it simple for now. My proposal would be to iterate over banks first and then ports on both sides (this is how it is currently implemented in C++, not sure about other targets). If the modified connection `-x->` is used, we iterate over banks first on the left and over ports first on the right. I think this should cover all cases of this pattern that I have seen so far. Of course, the precise syntax is up for discussion.

**Comment by edwardalee** [comment]
Yes, I agree. Let's go with the simpler solution. Reasoning about these connections will already be hard enough...

**Comment by cmnrd** [comment]
I opened a PR with a first implementation. I wonder if such connections should also get their own visual representation.

**Comment by Soroosh129** [comment]
I wonder if it makes sense to kill two birds with one stone and introduce an `offset` syntax:
```
nodes.out -> nodes.in + 1;
```
or 
```
nodes.out -> nodes.in offset 1;
```

In this case, you would not even need to do this:
```
out[(bank_index + 1) % num_nodes].set(bank_index);
```

The offsetting would be useful if for example you have a database replication in which two adjacent replicates are on the same server rack. So, a reaction in charge of backup might need to send an update to a bank member two indexes higher.

**Comment by cmnrd** [comment]
I don't think that offsetting would help here. It seems like an entirely different problem to me. I will prepare a few diagrams to make the problem more clear.

**Comment by cmnrd** [comment]
I think I can summarize yesterday's discussion at the weekly LF meeting as follows: We will implement the proposed feature as it solves the problem, but we should come up with a better syntax, and we should try to generalize this in the future to allow arbitrary connection patterns.

I was thinking about this overnight and believe that a function like syntax could work quite well. For instance we could use something like this:
```
nodes.out -> cross(nodes.in);
```
or
```
nodes.out -> interleaved(nodes.in);
```
Maybe someone has a name suggestion that is a little more clear. Also the validator could make sure that the `cross` or `interleave` operation is only used on multiports within a bank, as otherwise its behavior is not clear. This syntax would allow us to define arbitrary new patterns in the future. For instance, we could define `reversed(multiport)` if that is needed or introduce a general solution `map(multiport, pattern)`.

What needs to be known to the user is that the left hand side and the right hand side of the `->` operator are interpreted as lists of ports. The additional operators can change the order within those lists. This is particularly important in the case of banks with multiports, where we actually have a 2-dimensional array of ports that needs to be flattened, but it is not clear how the flattening needs to be done.

I am also back tracking a little on my argument about using the modifier only on the connection and not on individual ports. If we consider the above example and a bank `foo` with single port `in` and a multiport `bar`, then it is not really clear to me how
```
nodes.out -x-> foo.in, nodes.in, bar; 
```
would work. I think
```
nodes.out -> foo.in, interleaved(nodes.in), bar; 
``` 
is actually a bit clearer in expressing the intention of the programmer, although it is still non-obvious how the connection works. I think we should simply refrain from using such complex patterns (where possible).

What do you think?

**Comment by edwardalee** [comment]
I realized the same thing, also overnight (around 3AM :-) ).  The interleaving is a property not of a connection, but rather of an individual port reference of the form `b.p` where `b` is a bank and `p` is a multiport. Suppose `b` and `p` each have width 2.  Then the default ordering is `[b0.p0, b0.p1, b1.p0, b1.p1]`.  If you write `interleaved(b.p)` instead, then you get `[b0.p0, b1.p0, b0.p1, b1.p1]`.  I think this also makes clear that `interleaved` is a good description of this.

As a side note, I have this working in C with the previous syntax `-x->` and will push that as soon as I've replicated the `Cpp` tests.
